pallets/proofs-dealer/src/utils.rs

```diff
@@ -182,22 +181,60 @@ where
         // the current block. If there has been, the Provider should have included proofs for the
         // challenges in that block.
         let last_checkpoint_block = LastCheckpointBlock::<T>::get();
+        let mut checkpoint_challenges = None;
         if last_block_proven < last_checkpoint_block {
             // Add challenges from the Checkpoint Challenge block.
-            let checkpoint_challenges =
-                expect_or_err!(
+            checkpoint_challenges =
+                Some(expect_or_err!(
                     BlockToCheckpointChallenges::<T>::get(last_checkpoint_block),
                     "Checkpoint challenges not found, when dereferencing in last registered checkpoint challenge block.",
                     Error::<T>::CheckpointChallengesNotFound
-                );
-            challenges.extend(checkpoint_challenges);
+                ));
+
+            if let Some(ref checkpoint_challenges) = checkpoint_challenges {
+                challenges.extend(checkpoint_challenges.iter().map(|(key, _)| key));
+            }
         }
 
         // Verify forest proof.
-        let forest_keys_proven =
+        let mut forest_keys_proven =
             ForestVerifierFor::<T>::verify_proof(&root, &challenges, forest_proof)
                 .map_err(|_| Error::<T>::ForestProofVerificationFailed)?;
 
+        // Apply the delta to the Forest root for all mutations that are in checkpoint challenges.
+        if let Some(challenges) = checkpoint_challenges {
+            // Aggregate all mutations to apply to the Forest root.
+            let mutations: Vec<_> = challenges
+                .iter()
+                .filter_map(|(key, mutation)| match mutation {
+                    Some(Mutation::Remove) if forest_keys_proven.contains(key) => {
+                        Some((*key, Mutation::Remove))
+                    }
+                    Some(Mutation::Remove) | Some(Mutation::Add) | None => None,
```

Comment:
Technically, it shouldn't be possible to have a key with `Remove` mutation, that IS NOT in `forest_keys_proven`. Because that would mean that a checkpoint challenge has not been verified, therefore the verification should have failed. Basically this is an impossible state, and we should leave that clear in the code. I suggest something similar to what we do in `expect_or_err`, but instead of erroring out, we just return `None` (but in a test build, we panic).

Reply:
Modified the types and storage for the proofs dealer pallet to only accept RemoveMutation for checkpoint challenges.